onnxruntime/core/optimizer/constant_folding.cc

```diff
@@ -12,104 +13,156 @@ using namespace onnxruntime::common;
-        // by the Recurse call above
-        node->ContainsSubgraph() || !graph_utils::AllNodeInputsAreConstant(graph, *node, constant_inputs)) {
-      continue;
-    }
-
-    // override the EP while setting up OptimizerExecutionFrame::Info so that it will use the CPU kernel for Compute.
-    if (!cpu_ep) {
-      node->SetExecutionProviderType(kCpuExecutionProvider);
-    }
-
-    // Create execution frame for executing constant nodes.
-    OptimizerExecutionFrame::Info info({node}, constant_inputs);
-
-    // undo the EP change in case something fails prior to node removal
-    if (!cpu_ep) {
-      node->SetExecutionProviderType(ep_type);
-    }
+    // Resolve graph to force run shape and type inference.
+    // TODO: Resolve() is a heavy call. Better expose PerformTypeAndShapeInferencing() as one of graph APIs and call it here.
+    ORT_ENFORCE(graph.Resolve().IsOK());
 
-    std::vector<int> fetch_mlvalue_idxs;
-    for (const auto* node_out : node->OutputDefs()) {
-      fetch_mlvalue_idxs.push_back(info.GetMLValueIndex(node_out->Name()));
-    }
-
-    OptimizerExecutionFrame frame(info, fetch_mlvalue_idxs);
-
-    auto* kernel = info.GetKernel(node->Index());
-    if (kernel == nullptr)
-      continue;
-    OpKernelContext op_kernel_context(&frame, kernel, nullptr, logger);
-
-    ORT_RETURN_IF_ERROR(kernel->Compute(&op_kernel_context));
-
-    std::vector<OrtValue> fetches;
-    ORT_RETURN_IF_ERROR(frame.GetOutputs(fetches));
-
-    // Go over all output node args and substitute them with the newly computed tensors, which will be
-    // added to the graph as initializers.
-    ORT_ENFORCE(fetches.size() == node->OutputDefs().size());
-    bool unsupported_output_type = false;
-    for (size_t fetch_idx = 0; fetch_idx < fetches.size(); ++fetch_idx) {
-      OrtValue& ort_value = fetches[fetch_idx];
+    GraphViewer graph_viewer(graph);
+    auto& order = graph_viewer.GetNodesInTopologicalOrder();
 
-      if (!ort_value.IsTensor()) {
-        LOGS(logger, WARNING) << "Unsupported output type of " << ort_value.Type()
-                              << ". Can't constant fold " << node->OpType() << " node '" << node->Name() << "'";
-        unsupported_output_type = true;
-        break;
+    for (NodeIndex i : order) {
+      auto* node = graph.GetNode(i);
+      if (!node) {
+        continue;
       }
 
-      // Build the TensorProto that corresponds to the computed OrtValue and add it as initializer to the graph.
-      const auto* constant_arg_out = node->OutputDefs()[fetch_idx];
-      ORT_ENFORCE(ort_value.IsTensor());
-      const Tensor& out_tensor = ort_value.Get<Tensor>();
-      ONNX_NAMESPACE::TensorProto out_tensorproto = utils::TensorToTensorProto(out_tensor, constant_arg_out->Name());
+      ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level, logger));
+
+      bool convert_to_constant = false;
+      if (node->OpType().compare("Shape") == 0) {
+        auto shape = node->MutableInputDefs()[0]->Shape();
```

Comment:
Not quite understanding why does Shape need to be special cased. Why doesn't the constant folding work normally on a Shape node? If it's because of the checkon AllNodeInputsAreConstant would it be better to instead change the condition there rather than adding a separate branch for Shape here?

Reply:
For other nodes other than Shape, the computation is based on the real input tensors (they are all constants). But for Shape node, the input tensor is not constant, so the constant_inputs in line "OptimizerExecutionFrame::Info info({node}, constant_inputs);" cannot be prepared for the computation. But if the shape inference can infer the input's shape, we then can use it directly as the Shape node's output.